### PR TITLE
[CINFRA] Ignore ParticipantGone

### PR DESCRIPTION
**Pull request by apetenchea**
### Scope & Purpose

There are a some cases when "ParticipantGone" can be ignored. In `LogLeader.cpp`, this caused an exception to be thrown in a `noexcept` function.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### CODE REVIEW

File: arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp [hunk 1]
```diff
@@ -116,7 +116,18 @@ void CompactionManager::triggerAsyncCompaction(
...
+
+      auto storeRes = basics::catchToResultT([&]() -> auto {
+        return guard->storage.transaction();
+      });
+      if (storeRes.fail()) {
+        // The participant might be gone
+        LOG_CTX("d7724", DEBUG, self->loggerContext)
+            << "error during compaction: " << storeRes.result();
+        break;
+      }
```
Comment (goedderz) [inline arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp:128]: You should report the error, roughly as follows:
```suggestion
      if (storeRes.fail()) {
        // The participant might be gone
        LOG_CTX("d7724", ERR, self->loggerContext)
            << "error during compaction: " << storeRes.result();
        cresult.error = std::move(storeRes).result().error();
        guard->status.lastCompaction->error = cresult.error;
        clearCompactionRunning.fire();
        break;
      }
```
Please double-check the code, it was written blindly.

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp [hunk 2]
```diff
@@ -312,6 +310,14 @@ void replicated_log::LogLeader::executeAppendEntriesRequests(
...
                           return {};
                         });
+                    if (responseResult.fail()) {
+                      LOG_TOPIC("a32de", DEBUG, Logger::REPLICATION2)
+                          << "appendEntries failed, follower already gone: "
+                          << responseResult.result();
+                    }
+
+                    auto [preparedRequests, resolvedPromises] =
+                        std::move(responseResult.get());
```
Comment (goedderz) [inline arangod/Replication2/ReplicatedLog/LogLeader.cpp:320]: I don't think it's okay to silently stop here - the leader will stop working in a situation where it is not expected to. I rather think we should explicitly ignore resigned errors, but continue to crash with everything else.

**Review by goedderz — CHANGES_REQUESTED** [review]

File: arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp (hunk 1 above)
Comment (apetenchea) [reply arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp]: Addressed in https://github.com/arangodb/arangodb/pull/20628/commits/444c825fbff19e5f971ef36f315a0c7c2003a184

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp (hunk 2 above)
Comment (apetenchea) [reply arangod/Replication2/ReplicatedLog/LogLeader.cpp]: Addressed in https://github.com/arangodb/arangodb/pull/20628/commits/5b019267229a77636c84ae0a2b0e65035f7c3c36

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp [hunk 3]
```diff
@@ -273,69 +273,80 @@ void replicated_log::LogLeader::executeAppendEntriesRequests(
...
+                            << "received response from follower but leader "
+                               "already resigned, messageId = "
+                            << messageId;
+                      }
+                      return {};
+                    });
+                if (responseResult.fail()) {
+                  LOG_TOPIC("a32de", DEBUG, Logger::REPLICATION2)
+                      << "appendEntries failed, follower already gone: "
+                      << responseResult.result();
```
Comment (goedderz) [inline arangod/Replication2/ReplicatedLog/LogLeader.cpp:317]: The log message might not be accurate, unless you move it into the `if` below

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp [hunk 4]
```diff
@@ -273,69 +273,80 @@ void replicated_log::LogLeader::executeAppendEntriesRequests(
...
+                          TRI_ERROR_REPLICATION_REPLICATED_LOG_PARTICIPANT_GONE)) {
+                    // Ignore resign errors
+                    return;
                   }
-                });
+
+                  TRI_ASSERT(false)
+                      << "appendEntries failed, follower already gone: "
+                      << responseResult.result();
+                  FATAL_ERROR_EXIT();
```
Comment (goedderz) [inline arangod/Replication2/ReplicatedLog/LogLeader.cpp:328]:
```suggestion
                  ADB_PROD_ASSERT(responseResult.ok())
                      << "appendEntries failed with an unhandled error: "
                      << responseResult.result();
```

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp (hunk 3 above)
Comment (apetenchea) [reply arangod/Replication2/ReplicatedLog/LogLeader.cpp]: Good catch, addressed in https://github.com/arangodb/arangodb/pull/20628/commits/7f9c83b10f065a508acf68dc96cb815bd452dcd0

File: arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp [hunk 5]
```diff
@@ -126,16 +126,19 @@ void CompactionManager::triggerAsyncCompaction(
...
-                    CompactionStopReason::StorageManagerNotReady{}},
-            .compactedRange = {}};
-        guard->status.lastCompaction->error = cresult.error;
+
+        result::Error err{storeRes.result().errorNumber(),
+                          storeRes.result().errorMessage()};
+        auto& compaction = guard->status.inProgress.emplace();
+        guard->status.lastCompaction = guard->status.inProgress;
+        guard->status.inProgress.reset();
+        compaction.error = std::move(err);
```
Comment (goedderz) [inline arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp:135]: This looks a little convoluted, maybe a c&p relic? Why do you emplace `inProgress`, then copy it to `lastCompaction`, then reset it again, and _then_ write the `.error` field? Isn't this writing in the middle of a  freshly reset optional?

The way I see it: If `inProgress` is there, use that. If it's not, construct a new one, but don't forget to set the `.time` field to `CompactionStatus::clock::now()`. E.g.
```suggestion
        if (guard->status.inProgress.has_value()) {
          guard->status.lastCompaction = guard->status.inProgress;
          guard->status.inProgress.reset();
        } else {
          guard->status.lastCompaction = Compaction();
          guard->status.lastCompaction.time = CompactionStatus::clock::now();
        }
        guard->status.lastCompaction.error = std::move(err);
```
(which suggests that now `Compaction::range` should be an optional, but let's keep it at that ;).

Comment (apetenchea) [reply arangod/Replication2/ReplicatedLog/Components/CompactionManager.cpp]: Thanks, I simplified it in https://github.com/arangodb/arangodb/pull/20628/commits/f480fa0c1aa8c6cb6615763e7aafc7bf2e60b58d

**Pull request merged**